rhombus/rhombus/scribblings/ref-rx-charset.scrbl

```diff
@@ -0,0 +1,378 @@
+@rhombus(#%brackets, ~at rhombus/rx) operator). A character set
+represents a set of @rhombus(Char)s, but as long as the characters range
+in Unicode value from 0 to 255, a character set can be used as a set of
+bytes to match for a byte-mode @tech{regexp}.
+
+@doc(
+  rx_charset.macro '#%literal $string'
+  rx_charset.macro '#%literal $bytes'
+){
+
+ A literal @tech{string} or @tech{byte string} can be used as a character set.
+ Each character or byte is part of the set.
+
+@examples(
+  ~eval: rx_eval
+  rx'["a"]'.is_match("a")
+  rx'["a"]'.is_match("b")
+  rx'["abc"]'.is_match("b")
+)
+
+}
+
+
+@doc(
+  rx_charset.macro '$charset #%juxtapose $charset'
+  rx_charset.macro '$charset || $charset'
+  rx_charset.macro '$charset #%call ($charset)'
+){
+
+ Character sets that are adjacent or joined with
+ @rhombus(||, ~at rhombus/rx_charset) form a larger character set that
+ includes all combined elements, i.e., a union of the sets. An implicit
+ @rhombus(#%call, ~at rhombus/rx_charset) form is treated like
+ @rhombus(#%juxtapose, ~at rhombus/rx_charset), consistent with implicit uses of
+ parentheses for grouping as handled by
+ @rhombus(#%parens, ~at rhombus/rx_charset).
+
+@examples(
+  ~eval: rx_eval
+  rx'["a" "b"]'.is_match("a")
+  rx'["a" "b"]'.is_match("b")
+  rx'["a" "b"]'.is_match("c")
+  rx'["a" || "b"]'.is_match("b")
+)
+
+}
+
+
+@doc(
+  rx_charset.macro '#%parens ($charset)'
+){
+
+ A parenthesized character set is equivalent to the @rhombus(charset)
+ inside the parentheses. That is, parentheses are just for grouping and
+ resolving precedence mismatches.
+
+@examples(
+  ~eval: rx_eval
+  rx'["a" "b" "c"]'.is_match("a")
+  rx'[("a" "b") "c"]'.is_match("a")
+)
+
+}
+
+
+@doc(
+  rx_charset.macro '$charset - $charset'
+){
+
+ Assuming that each @rhombus(charset) contains a single character,
+ creates a charset that has those two characters and all characters in
+ between (based on @rhombus(Char.to_int) values). An error is reported if
+ either @rhombus(charset) has zero or multiple characters.
+
+@examples(
+  ~eval: rx_eval
+  rx'["a" - "y"]'.is_match("a")
+  rx'["a" - "y"]'.is_match("x")
+  rx'["a" - "y"]'.is_match("z")
+)
+
+}
+
+
+@doc(
+  rx_charset.macro '$charset && $charset'
+){
+
+ Creates a character set that has each character in both the first
+ @rhombus(charset) and the second @rhombus(charset), i.e., an
+ intersection of the sets.
+
+@examples(
+  ~eval: rx_eval
+  rx'[("a" - "f") && ("c" - "h")]'.is_match("a")
+  rx'[("a" - "f") && ("c" - "h")]'.is_match("d")
+)
+
+}
+
+
+
+@doc(
+  rx_charset.macro '$charset -- $charset'
+){
+
+ Creates a character set that starts with the character of the first
+ @rhombus(charset) and removes each character of the second
+ @rhombus(charset), i.e., set difference.
+
+@examples(
+  ~eval: rx_eval
+  rx'[("a" - "z") -- ("m" - "p")]'.is_match("n")
+  rx'[("a" - "z") -- ("m" - "p")]'.is_match("a")
+)
+
+}
+
+
+@doc(
+  rx_charset.macro '^ $charset'
+  rx_charset.macro '! $charset'
```

Comment:
Why are there multiple characters used for the same purpose? I think it's a waste of characters. And they are not syntactically meaningfully different.

Reply:
I'd be in favor of picking one, but I wasn't sure whether `^` or `!` is better. I lean toward `!`, and I lean a little more in that direction if we also drop `^` as a anchor pattern.